File: ios-cuantrack/CuanTrack/Models/Budget.swift
//
//  Budget.swift
//  CuanTrack
//

import Foundation

struct BudgetCategory: Identifiable, Codable {
    let id: UUID
    var category: TransactionCategory
    var monthlyLimit: Double
    var month: Int
    var year: Int
    var spent: Double = 0

    init(
        id: UUID = UUID(),
        category: TransactionCategory,
        monthlyLimit: Double = 0,
        month: Int? = nil,
        year: Int? = nil
    ) {
        self.id = id
        self.category = category
        self.monthlyLimit = monthlyLimit
        let now = Date()
        let calendar = Calendar.current
        self.month = month ?? calendar.component(.month, from: now)
        self.year = year ?? calendar.component(.year, from: now)
    }

    var formattedLimit: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.groupingSeparator = "."
        formatter.locale = Locale(identifier: "id_ID")
        return "Rp \(formatter.string(from: NSNumber(value: monthlyLimit)) ?? "0")"
    }

    var percentSpent: Double {
        guard monthlyLimit > 0 else { return 0 }
        return min(spent / monthlyLimit, 1.0)
    }
}


enum BudgetHealth {
    case safe
    case warning
    case critical

    var label: String {
        switch self {
        case .safe: "Aman"
        case .warning: "Hampir habis"
        case .critical: "Melebihi"
        }
    }

    static func from(percent: Double) -> BudgetHealth {
        if percent >= 1.0 { return .critical }
        if percent >= 0.75 { return .warning }
        return .safe
    }
}

File: android-cuantrack/app/src/main/java/com/rork/cuantrackandroid/models/Budget.kt
package com.rork.cuantrackandroid.models

import androidx.compose.ui.graphics.Color
import java.util.UUID

data class BudgetCategory(
    val id: UUID = UUID.randomUUID(),
    val category: TransactionCategory = TransactionCategory.OTHER,
    val monthlyLimit: Double = 0.0,
    val spent: Double = 0.0,
    val month: Int = 0,
    val year: Int = 0
) {
    val formattedLimit: String
        get() {
            val nf = java.text.NumberFormat.getNumberInstance(java.util.Locale("id", "ID"))
            nf.maximumFractionDigits = 0
            return "Rp ${nf.format(monthlyLimit)}"
        }

    val percentSpent: Float
        get() = if (monthlyLimit > 0) (spent / monthlyLimit).toFloat().coerceIn(0f, 1f) else 0f

    val health: BudgetHealth
        get() = BudgetHealth.fromPercent(percentSpent.toDouble())

    val healthLabel: String
        get() = health.label

    val progressColor: Color
        get() = when (health) {
            BudgetHealth.SAFE -> category.color
            BudgetHealth.WARNING -> Color(0xFFF59E0B)
            BudgetHealth.CRITICAL -> Color(0xFFDC2626)
        }
}

enum class BudgetHealth(val label: String) {
    SAFE("Aman"),
    WARNING("Hampir habis"),
    CRITICAL("Melebihi");

    companion object {
        fun fromPercent(percent: Double): BudgetHealth = when {
            percent >= 1.0 -> CRITICAL
            percent >= 0.75 -> WARNING
            else -> SAFE
        }
    }
}
